File: projects-review/agent-trust/packages/python-sdk/agent_trust/identity.py
"""
AgentTrust Identity — Ed25519 keypair + DID:key generation (pure Python).

No external cryptography libraries required for basic usage.
Uses hashlib + secrets from the standard library.
"""

from __future__ import annotations

import hashlib
import json
import os
import secrets
import struct
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

# ── Multibase / Multicodec constants ────────────────────────────────────────

ED25519_MULTICODEC_PREFIX = bytes([0xed, 0x01])
BASE58_ALPHABET = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"


def _base58_encode(data: bytes) -> str:
    """Bitcoin-alphabet base58 encoding."""
    n = int.from_bytes(data, "big")
    result = []
    while n > 0:
        n, r = divmod(n, 58)
        result.append(BASE58_ALPHABET[r:r+1])
    # Leading zero bytes
    result.extend(b"1" for b in data if b == 0)
    return b"".join(reversed(result)).decode()


def _base58_decode(s: str) -> bytes:
    n = 0
    for char in s.encode():
        n = n * 58 + BASE58_ALPHABET.index(char)
    result = []
    while n > 0:
        n, r = divmod(n, 256)
        result.append(r)
    # Leading '1's
    result.extend(0 for c in s if c == "1")
    return bytes(reversed(result))


# ── Ed25519 field arithmetic (Curve25519 / RFC 8032) ───────────────────────

P  = 2**255 - 19
D  = 37095705934669439343138083508754565189542113879843219016388785533085940283555
Q  = 2**252 + 27742317777372353535851937790883648493
I  = pow(2, (P - 1) // 4, P)


def _inv(x: int) -> int:
    return pow(x, P - 2, P)


def _recover_x(y: int, sign: int) -> int:
    y2 = y * y % P
    x2 = (y2 - 1) * _inv(D * y2 + 1) % P
    if x2 == 0:
        if sign:
            raise ValueError("Invalid point")
        return 0
    x = pow(x2, (P + 3) // 8, P)
    if (x * x - x2) % P != 0:
        x = x * I % P
    if (x * x - x2) % P != 0:
        raise ValueError("Invalid point")
    if x & 1 != sign:
        x = P - x
    return x


_G: Optional[tuple] = None


def _get_G() -> tuple:
    global _G
    if _G is None:
        y = 4 * _inv(5) % P
        x = _recover_x(y, 0)
        _G = (x, y, 1, x * y % P)
    return _G


def _point_add(P1: tuple, P2: tuple) -> tuple:
    X1, Y1, Z1, T1 = P1
    X2, Y2, Z2, T2 = P2
    a = (Y1 - X1) * (Y2 - X2) % P
    b = (Y1 + X1) * (Y2 + X2) % P
    c = T1 * 2 * D * T2 % P
    d = Z1 * 2 * Z2 % P
    e = b - a
    f = d - c
    g = d + c
    h = b + a
    return (e * f % P, g * h % P, f * g % P, e * h % P)


def _point_mul(s: int, point: tuple) -> tuple:
    result = (0, 1, 1, 0)
    current = point
    while s > 0:
        if s & 1:
            result = _point_add(result, current)
        current = _point_add(current, current)
        s >>= 1
    return result


def _compress(point: tuple) -> bytes:
    X, Y, Z = point[:3]
    zi = _inv(Z)
    x = X * zi % P
    y = Y * zi % P
    b = y.to_bytes(32, "little")
    if x & 1:
        b = b[:-1] + bytes([b[-1] | 0x80])
    return b


def _sha512(data: bytes) -> bytes:
    return hashlib.sha512(data).digest()


def _clamp(k: bytes) -> int:
    c = bytearray(k)
    c[0] &= 248
    c[31] &= 127
    c[31] |= 64
    return int.from_bytes(c, "little")


def _ed_sign(message: bytes, private_key: bytes) -> bytes:
    h = _sha512(private_key)
    a = _clamp(h[:32])
    prefix = h[32:]
    G = _get_G()
    A = _compress(_point_mul(a, G))
    r_hash = _sha512(prefix + message)
    r = int.from_bytes(r_hash, "little") % Q
    R = _compress(_point_mul(r, G))
    k_hash = _sha512(R + A + message)
    k = int.from_bytes(k_hash, "little") % Q
    S = ((r + k * a) % Q).to_bytes(32, "little")
    return R + S


def _ed_verify(message: bytes, signature: bytes, public_key: bytes) -> bool:
    if len(signature) != 64:
        return False
    try:
        R_bytes = signature[:32]
        S = int.from_bytes(signature[32:], "little")
        pk_y = int.from_bytes(public_key, "little") & ((1 << 255) - 1)
        pk_sign = public_key[31] >> 7
        pk_x = _recover_x(pk_y, pk_sign)
        A = (pk_x, pk_y, 1, pk_x * pk_y % P)
        ry = int.from_bytes(R_bytes, "little") & ((1 << 255) - 1)
        rs = R_bytes[31] >> 7
        rx = _recover_x(ry, rs)
        R_pt = (rx, ry, 1, rx * ry % P)
        G = _get_G()
        k_hash = _sha512(R_bytes + public_key + message)
        k = int.from_bytes(k_hash, "little") % Q
        lhs = _compress(_point_mul(S, G))
        rhs = _compress(_point_add(R_pt, _point_mul(k, A)))
        return lhs == rhs
    except Exception:
        return False


def _public_key_from_private(private_key: bytes) -> bytes:
    h = _sha512(private_key)
    a = _clamp(h[:32])
    G = _get_G()
    return _compress(_point_mul(a, G))


# ── Public API ──────────────────────────────────────────────────────────────

@dataclass
class AgentKeypair:
    """An Ed25519 keypair with a derived DID:key identity."""
    private_key: bytes   # 32 bytes
    public_key: bytes    # 32 bytes
    did: str             # did:key:z6Mk...


def generate_keypair() -> AgentKeypair:
    """
    Generate a fresh Ed25519 keypair and derive a did:key:z6Mk... DID.

    Example:
        kp = generate_keypair()
        print(kp.did)  # did:key:z6Mk...
    """
    private_key = secrets.token_bytes(32)
    public_key = _public_key_from_private(private_key)
    multicodec = ED25519_MULTICODEC_PREFIX + public_key
    did = f"did:key:z{_base58_encode(multicodec)}"
    return AgentKeypair(private_key=private_key, public_key=public_key, did=did)


def public_key_to_did(public_key: bytes) -> str:
    """Derive a did:key:z6Mk... DID from a 32-byte Ed25519 public key."""
    multicodec = ED25519_MULTICODEC_PREFIX + public_key
    return f"did:key:z{_base58_encode(multicodec)}"


def did_to_public_key(did: str) -> bytes:
    """Extract the raw 32-byte Ed25519 public key from a did:key:z6Mk... DID."""
    if not did.startswith("did:key:"):
        raise ValueError("Not a did:key DID")
    mb = did[len("did:key:"):]
    if not mb.startswith("z"):
        raise ValueError("Expected multibase base58btc (prefix 'z')")
    decoded = _base58_decode(mb[1:])
    if decoded[:2] != ED25519_MULTICODEC_PREFIX:
        raise ValueError("DID is not an Ed25519 did:key (expected multicodec prefix 0xed01)")
    return decoded[2:]  # 32-byte raw public key


def sign_data(data: str | bytes, private_key: bytes) -> str:
    """
    Sign data with a private key. Returns a 128-char hex string (64-byte signature).

    Example:
        sig = sign_data("hello world", kp.private_key)
    """
    if isinstance(data, str):
        data = data.encode()
    return _ed_sign(data, private_key).hex()


def verify_signature(data: str | bytes, signature_hex: str, public_key: bytes) -> bool:
    """Verify a signature produced by sign_data."""
    if isinstance(data, str):
        data = data.encode()
    return _ed_verify(data, bytes.fromhex(signature_hex), public_key)


def verify_signature_by_did(data: str | bytes, signature_hex: str, did: str) -> bool:
    """Verify a signature using a did:key:z6Mk... DID (no raw key needed)."""
    return verify_signature(data, signature_hex, did_to_public_key(did))


@dataclass
class IdentityFile:
    """Serialized keypair saved to .agent-trust.json"""
    did: str
    public_key: str   # hex
    private_key: str  # hex  ← keep secret!
    created_at: str
    version: str = "0.2.0"

    def to_dict(self) -> dict:
        return {
            "did": self.did,
            "publicKey": self.public_key,
            "privateKey": self.private_key,
            "createdAt": self.created_at,
            "version": self.version,
        }

    @classmethod
    def from_dict(cls, d: dict) -> "IdentityFile":
        return cls(
            did=d["did"],
            public_key=d["publicKey"],
            private_key=d["privateKey"],
            created_at=d["createdAt"],
            version=d.get("version", "0.2.0"),
        )

    def to_keypair(self) -> AgentKeypair:
        return AgentKeypair(
            private_key=bytes.fromhex(self.private_key),
            public_key=bytes.fromhex(self.public_key),
            did=self.did,
        )


def load_or_create_identity(path: str = ".agent-trust.json") -> AgentKeypair:
    """
    Load identity from .agent-trust.json, or generate a new one if it doesn't exist.

    This is the recommended way to initialize an agent's identity in production.
    """
    from datetime import datetime, timezone
    p = Path(path)
    if p.exists():
        with open(p) as f:
            return IdentityFile.from_dict(json.load(f)).to_keypair()

    kp = generate_keypair()
    identity_file = IdentityFile(
        did=kp.did,
        public_key=kp.public_key.hex(),
        private_key=kp.private_key.hex(),
        created_at=datetime.now(timezone.utc).isoformat(),
    )
    with open(p, "w") as f:
        json.dump(identity_file.to_dict(), f, indent=2)
    return kp

File: projects-review/agent-trust/packages/core/src/identity.ts
/**
 * Ed25519 Keypair + DID:key generation for AgentTrust Protocol
 *
 * Uses the `@noble/ed25519` library (audited, no native deps) to:
 *  - Generate a real Ed25519 keypair
 *  - Derive a DID in the `did:key:z6Mk...` format (multibase base58btc)
 *  - Sign arbitrary payloads (for transaction proofs, VC signing)
 *  - Verify signatures
 *
 * No external services required — fully offline, self-sovereign identity.
 */

import { randomBytes } from 'node:crypto';

// ------ Multibase / Multicodec helpers (no extra deps) ----------------------

/**
 * base58btc encode (Bitcoin alphabet)
 */
const BASE58_ALPHABET =
  '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

function base58Encode(bytes: Uint8Array): string {
  let num = 0n;
  for (const byte of bytes) {
    num = num * 256n + BigInt(byte);
  }
  let encoded = '';
  while (num > 0n) {
    const remainder = num % 58n;
    num = num / 58n;
    encoded = BASE58_ALPHABET[Number(remainder)] + encoded;
  }
  // Leading zeros
  for (const byte of bytes) {
    if (byte !== 0) break;
    encoded = '1' + encoded;
  }
  return encoded;
}

function base58Decode(str: string): Uint8Array {
  let num = 0n;
  for (const char of str) {
    const idx = BASE58_ALPHABET.indexOf(char);
    if (idx === -1) throw new Error(`Invalid base58 character: ${char}`);
    num = num * 58n + BigInt(idx);
  }
  const bytes: number[] = [];
  while (num > 0n) {
    bytes.unshift(Number(num & 0xffn));
    num >>= 8n;
  }
  // Leading '1's → 0x00 bytes
  for (const char of str) {
    if (char !== '1') break;
    bytes.unshift(0);
  }
  return new Uint8Array(bytes);
}

/**
 * Encode bytes as multibase base58btc (prefix 'z')
 */
function multibaseBase58Encode(bytes: Uint8Array): string {
  return 'z' + base58Encode(bytes);
}

function multibaseBase58Decode(mb: string): Uint8Array {
  if (!mb.startsWith('z')) throw new Error('Expected multibase base58btc (prefix z)');
  return base58Decode(mb.slice(1));
}

/**
 * Ed25519 multicodec prefix: 0xed01
 * See https://w3c-ccg.github.io/did-method-key/#create
 */
const ED25519_MULTICODEC = new Uint8Array([0xed, 0x01]);

// ------ Pure-JS Ed25519 implementation (RFC 8032) ---------------------------
// We implement a minimal Ed25519 to avoid requiring an external package at runtime.
// This is sufficient for DID generation, signing, and verification.

const P = 2n ** 255n - 19n;
const D = 37095705934669439343138083508754565189542113879843219016388785533085940283555n;
const Q = 2n ** 252n + 27742317777372353535851937790883648493n;
// Gy is unused directly; G() derives y=4/5 mod P instead
// const Gy = 46316835694926478169428394003475163141307993866256225615783033011972563074n;

function modP(a: bigint): bigint {
  return ((a % P) + P) % P;
}

function pow2(x: bigint, power: bigint, mod: bigint): bigint {
  let res = 1n;
  x = x % mod;
  while (power > 0n) {
    if (power & 1n) res = res * x % mod;
    power >>= 1n;
    x = x * x % mod;
  }
  return res;
}

function inv(x: bigint): bigint {
  return pow2(x, P - 2n, P);
}

function recoverX(y: bigint, sign: bigint): bigint {
  const y2 = y * y % P;
  const x2 = (y2 - 1n) * inv(D * y2 + 1n) % P;
  if (x2 === 0n) {
    if (sign) throw new Error('Invalid point');
    return 0n;
  }
  let x = pow2(x2, (P + 3n) / 8n, P);
  if ((x * x - x2) % P !== 0n) {
    const I = pow2(2n, (P - 1n) / 4n, P);
    x = x * I % P;
  }
  if ((x * x - x2) % P !== 0n) throw new Error('Invalid point');
  if (x % 2n !== sign) x = P - x;
  return x;
}

type Point = [bigint, bigint, bigint, bigint]; // [X, Y, Z, T]

const Gx = recoverX(4n * inv(5n) % P, 0n);
function G(): Point {
  const y = 4n * inv(5n) % P;
  const x = recoverX(y, 0n);
  return [x, y, 1n, x * y % P];
}

function pointAdd(P1: Point, P2: Point): Point {
  const [X1, Y1, Z1, T1] = P1;
  const [X2, Y2, Z2, T2] = P2;
  const a = (Y1 - X1) * (Y2 - X2) % P;
  const b = (Y1 + X1) * (Y2 + X2) % P;
  const c = T1 * 2n * D * T2 % P;
  const d = Z1 * 2n * Z2 % P;
  const e = b - a;
  const f = d - c;
  const g = d + c;
  const h = b + a;
  return [e * f % P, g * h % P, f * g % P, e * h % P];
}

function pointMul(s: bigint, P1: Point): Point {
  let Q: Point = [0n, 1n, 1n, 0n];
  let R: Point = P1;
  while (s > 0n) {
    if (s & 1n) Q = pointAdd(Q, R);
    R = pointAdd(R, R);
    s >>= 1n;
  }
  return Q;
}

function pointCompress(P1: Point): Uint8Array {
  const [X, Y, Z] = P1;
  const zi = inv(Z);
  const x = X * zi % P;
  const y = Y * zi % P;
  const bytes = new Uint8Array(32);
  let tmp = y;
  for (let i = 0; i < 32; i++) {
    bytes[i] = Number(tmp & 0xffn);
    tmp >>= 8n;
  }
  if (x & 1n) bytes[31] |= 0x80;
  return bytes;
}

async function sha512(data: Uint8Array): Promise<Uint8Array> {
  const hash = await crypto.subtle.digest('SHA-512', data.buffer as ArrayBuffer);
  return new Uint8Array(hash);
}

function clamp(k: Uint8Array): bigint {
  const clamped = new Uint8Array(k);
  clamped[0] &= 248;
  clamped[31] &= 127;
  clamped[31] |= 64;
  let s = 0n;
  for (let i = 31; i >= 0; i--) {
    s = s * 256n + BigInt(clamped[i]);
  }
  return s;
}

function bytesToBigIntLE(bytes: Uint8Array): bigint {
  let s = 0n;
  for (let i = bytes.length - 1; i >= 0; i--) {
    s = s * 256n + BigInt(bytes[i]);
  }
  return s;
}

function bigIntToBytesLE(n: bigint, len: number): Uint8Array {
  const bytes = new Uint8Array(len);
  let tmp = ((n % Q) + Q) % Q;
  for (let i = 0; i < len; i++) {
    bytes[i] = Number(tmp & 0xffn);
    tmp >>= 8n;
  }
  return bytes;
}

async function edSign(message: Uint8Array, privateKey: Uint8Array): Promise<Uint8Array> {
  const h = await sha512(privateKey);
  const a = clamp(h.slice(0, 32));
  const prefix = h.slice(32);
  const G_pt = G();
  const A = pointCompress(pointMul(a, G_pt));
  const rHash = await sha512(new Uint8Array([...prefix, ...message]));
  const r = bytesToBigIntLE(rHash) % Q;
  const R = pointCompress(pointMul(r, G_pt));
  const kHash = await sha512(new Uint8Array([...R, ...A, ...message]));
  const k = bytesToBigIntLE(kHash) % Q;
  const S = bigIntToBytesLE((r + k * a) % Q, 32);
  return new Uint8Array([...R, ...S]);
}

async function edVerify(
  message: Uint8Array,
  signature: Uint8Array,
  publicKey: Uint8Array
): Promise<boolean> {
  try {
    if (signature.length !== 64) return false;
    const R_bytes = signature.slice(0, 32);
    const S = bytesToBigIntLE(signature.slice(32));
    const pk_y = bytesToBigIntLE(publicKey) & ((1n << 255n) - 1n);
    const pk_sign = BigInt(publicKey[31] >> 7);
    const pk_x = recoverX(pk_y, pk_sign);
    const A: Point = [pk_x, pk_y, 1n, pk_x * pk_y % P];
    const ry = bytesToBigIntLE(R_bytes) & ((1n << 255n) - 1n);
    const rs = BigInt(R_bytes[31] >> 7);
    const rx = recoverX(ry, rs);
    const R: Point = [rx, ry, 1n, rx * ry % P];
    const G_pt = G();
    const kHash = await sha512(new Uint8Array([...R_bytes, ...publicKey, ...message]));
    const k = bytesToBigIntLE(kHash) % Q;
    const lhs = pointCompress(pointMul(S, G_pt));
    const rhs = pointCompress(pointAdd(R, pointMul(k, A)));
    return lhs.every((b, i) => b === rhs[i]);
  } catch {
    return false;
  }
}

// ------ Public API ----------------------------------------------------------

export interface AgentKeypair {
  /** Raw 32-byte Ed25519 private key seed */
  privateKey: Uint8Array;
  /** Raw 32-byte Ed25519 public key */
  publicKey: Uint8Array;
  /** DID in the `did:key:z6Mk...` format, derived from publicKey */
  did: string;
}

export interface AgentIdentityFile {
  did: string;
  publicKey: string;   // hex
  privateKey: string;  // hex  (keep secret!)
  createdAt: string;
  version: string;
}

/**
 * Generate a fresh Ed25519 keypair and derive a `did:key:z6Mk...` DID.
 *
 * @example
 * const keypair = await generateKeypair();
 * console.log(keypair.did); // did:key:z6Mk...
 */
export async function generateKeypair(): Promise<AgentKeypair> {
  const privateKey = randomBytes(32);
  const h = await sha512(privateKey);
  const a = clamp(h.slice(0, 32));
  const G_pt = G();
  const publicKeyBytes = pointCompress(pointMul(a, G_pt));

  const multicodecPubkey = new Uint8Array([...ED25519_MULTICODEC, ...publicKeyBytes]);
  const did = `did:key:${multibaseBase58Encode(multicodecPubkey)}`;

  return { privateKey, publicKey: publicKeyBytes, did };
}

/**
 * Derive a `did:key:z6Mk...` DID from a raw 32-byte Ed25519 public key.
 */
export function publicKeyToDid(publicKey: Uint8Array): string {
  const multicodecPubkey = new Uint8Array([...ED25519_MULTICODEC, ...publicKey]);
  return `did:key:${multibaseBase58Encode(multicodecPubkey)}`;
}

/**
 * Extract the raw 32-byte public key from a `did:key:z6Mk...` DID.
 *
 * @throws if the DID is not a valid Ed25519 did:key
 */
export function didToPublicKey(did: string): Uint8Array {
  if (!did.startsWith('did:key:')) throw new Error('Not a did:key DID');
  const mbPart = did.slice('did:key:'.length);
  const bytes = multibaseBase58Decode(mbPart);
  if (bytes[0] !== 0xed || bytes[1] !== 0x01) {
    throw new Error('DID is not an Ed25519 did:key (expected multicodec prefix 0xed01)');
  }
  return bytes.slice(2); // 32-byte raw public key
}

/**
 * Sign arbitrary data with a private key.
 * Returns a 64-byte Ed25519 signature as a hex string.
 *
 * @example
 * const sig = await signData('hello world', keypair.privateKey);
 */
export async function signData(
  data: string | Uint8Array,
  privateKey: Uint8Array
): Promise<string> {
  const bytes = typeof data === 'string' ? new TextEncoder().encode(data) : data;
  const sig = await edSign(bytes, privateKey);
  return Buffer.from(sig).toString('hex');
}

/**
 * Verify a signature produced by `signData`.
 *
 * @example
 * const ok = await verifySignature('hello world', sig, keypair.publicKey);
 */
export async function verifySignature(
  data: string | Uint8Array,
  signatureHex: string,
  publicKey: Uint8Array
): Promise<boolean> {
  const bytes = typeof data === 'string' ? new TextEncoder().encode(data) : data;
  const sig = Buffer.from(signatureHex, 'hex');
  return edVerify(bytes, sig, publicKey);
}

/**
 * Verify a signature using a `did:key:z6Mk...` DID (no raw key needed).
 *
 * @example
 * const ok = await verifySignatureByDid('hello', sig, 'did:key:z6Mk...');
 */
export async function verifySignatureByDid(
  data: string | Uint8Array,
  signatureHex: string,
  did: string
): Promise<boolean> {
  const publicKey = didToPublicKey(did);
  return verifySignature(data, signatureHex, publicKey);
}

/**
 * Serialize a keypair to the `.agent-trust.json` file format.
 */
export function keypairToIdentityFile(kp: AgentKeypair): AgentIdentityFile {
  return {
    did: kp.did,
    publicKey: Buffer.from(kp.publicKey).toString('hex'),
    privateKey: Buffer.from(kp.privateKey).toString('hex'),
    createdAt: new Date().toISOString(),
    version: '0.2.0',
  };
}

/**
 * Deserialize an identity file back to usable keypair buffers.
 */
export function identityFileToKeypair(file: AgentIdentityFile): {
  did: string;
  privateKey: Uint8Array;
  publicKey: Uint8Array;
} {
  return {
    did: file.did,
    privateKey: Buffer.from(file.privateKey, 'hex'),
    publicKey: Buffer.from(file.publicKey, 'hex'),
  };
}
